File: src/main/java/com/sslproxy/coordinator/oracle/JsonFields.java
package com.sslproxy.coordinator.oracle;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.time.OffsetDateTime;
import java.time.ZoneOffset;

final class JsonFields {

    private JsonFields() {}

    static long rowSequence(int index, String context) {
        return Math.addExact(index, 1);
    }

    static String rawJson(ObjectMapper objectMapper, JsonNode row, String context) {
        try {
            return objectMapper.writeValueAsString(row);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("encode raw " + context + " json: " + e.getMessage(), e);
        }
    }

    static String requiredString(JsonNode row, String field, String context) {
        return optionalString(row, field)
                .filter(value -> !value.isBlank())
                .orElseThrow(() -> new IllegalArgumentException(context + " row missing " + field));
    }

    static java.util.Optional<String> optionalString(JsonNode row, String field) {
        JsonNode value = row.get(field);
        if (value == null || value.isNull()) {
            return java.util.Optional.empty();
        }
        if (!value.isTextual()) {
            return java.util.Optional.empty();
        }
        String text = value.asText().trim();
        return text.isEmpty() ? java.util.Optional.empty() : java.util.Optional.of(text);
    }

    static String stringAlias(JsonNode row, String first, String second, String context) {
        return optionalString(row, first)
                .or(() -> optionalString(row, second))
                .filter(value -> !value.isBlank())
                .orElseThrow(() -> new IllegalArgumentException(context + " row missing " + first + "/" + second));
    }

    static long requiredLong(JsonNode row, String field, String context) {
        return optionalLong(row, field)
                .orElseThrow(() -> new IllegalArgumentException(context + " row missing numeric " + field));
    }

    static java.util.Optional<Long> optionalLong(JsonNode row, String field) {
        JsonNode value = row.get(field);
        if (value == null || value.isNull() || !value.isNumber()) {
            return java.util.Optional.empty();
        }
        return java.util.Optional.of(value.longValue());
    }

    static long longAlias(JsonNode row, String first, String second, String context) {
        return optionalLong(row, first)
                .or(() -> optionalLong(row, second))
                .orElseThrow(() -> new IllegalArgumentException(context + " row missing numeric " + first + "/" + second));
    }

    static java.util.Optional<Long> nestedLong(JsonNode row, String object, String field) {
        JsonNode parent = row.get(object);
        if (parent == null || parent.isNull()) {
            return java.util.Optional.empty();
        }
        return optionalLong(parent, field);
    }

    static java.util.Optional<Double> optionalDouble(JsonNode row, String field) {
        JsonNode value = row.get(field);
        if (value == null || value.isNull() || !value.isNumber()) {
            return java.util.Optional.empty();
        }
        return java.util.Optional.of(value.doubleValue());
    }

    static java.util.Optional<Double> nestedDouble(JsonNode row, String object, String field) {
        JsonNode parent = row.get(object);
        if (parent == null || parent.isNull()) {
            return java.util.Optional.empty();
        }
        return optionalDouble(parent, field);
    }

    static long boolFlag(JsonNode row, String field) {
        JsonNode value = row.get(field);
        return value != null && value.isBoolean() && value.asBoolean() ? 1L : 0L;
    }

    static OffsetDateTime requiredTimestamp(JsonNode row, String field, String context) {
        return parseTimestamp(requiredString(row, field, context), field);
    }

    static OffsetDateTime timestampAlias(JsonNode row, String first, String second, String context) {
        String value = optionalString(row, first)
                .or(() -> optionalString(row, second))
                .orElseThrow(() -> new IllegalArgumentException(context + " row missing " + first + "/" + second));
        return parseTimestamp(value, first);
    }

    static java.util.Optional<OffsetDateTime> optionalTimestamp(JsonNode row, String field) {
        return optionalString(row, field).map(value -> parseTimestamp(value, field));
    }

    static OffsetDateTime parseTimestamp(String value, String field) {
        try {
            return OffsetDateTime.parse(value).withOffsetSameInstant(ZoneOffset.UTC);
        } catch (RuntimeException e) {
            throw new IllegalArgumentException("decode " + field + " timestamp: " + e.getMessage(), e);
        }
    }

    static String jsonArrayString(ObjectMapper objectMapper, JsonNode row, String field) {
        JsonNode value = row.get(field);
        if (value == null || value.isNull()) {
            return "[]";
        }
        if (value.isTextual()) {
            return value.asText();
        }
        if (!value.isArray()) {
            throw new IllegalArgumentException(field + " must be a JSON array or string");
        }
        try {
            return objectMapper.writeValueAsString(value);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("encode " + field + ": " + e.getMessage(), e);
        }
    }
}

File: src/main/scala/com/sslproxy/coordinator/tidb/JsonFields.scala
package com.sslproxy.coordinator.tidb

import io.circe.Json
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import scala.util.Try

/** Circe-based JSON field extraction helpers. Ported from JsonFields.java. */
object JsonFields:

  private val timestampParser: DateTimeFormatter =
    DateTimeFormatter.ISO_OFFSET_DATE_TIME

  def requiredString(row: Json, field: String, context: String): String =
    row.hcursor.get[String](field) match
      case Right(value) => value
      case Left(_) =>
        throw new IllegalArgumentException(
          s"Missing required string field '$field' in $context"
        )

  def optionalString(row: Json, field: String): Option[String] =
    row.hcursor.get[String](field).toOption.filter(_.nonEmpty)

  def requiredLong(row: Json, field: String, context: String): Long =
    row.hcursor.get[Long](field) match
      case Right(value) => value
      case Left(_) =>
        throw new IllegalArgumentException(
          s"Missing required long field '$field' in $context"
        )

  def optionalLong(row: Json, field: String): Option[Long] =
    row.hcursor.get[Long](field).toOption

  def optionalDouble(row: Json, field: String): Option[Double] =
    row.hcursor.get[Double](field).toOption

  def requiredTimestamp(row: Json, field: String, context: String): OffsetDateTime =
    row.hcursor.get[String](field) match
      case Right(value) =>
        parseTimestamp(value).getOrElse(
          throw new IllegalArgumentException(
            s"Invalid timestamp '$value' for field '$field' in $context"
          )
        )
      case Left(_) =>
        throw new IllegalArgumentException(
          s"Missing required timestamp field '$field' in $context"
        )

  def optionalTimestamp(row: Json, field: String): Option[OffsetDateTime] =
    row.hcursor.get[String](field).toOption.flatMap(parseTimestamp)

  /** Alias: try primary field, fall back to secondary. */
  def timestampAlias(
      row: Json,
      primary: String,
      secondary: String,
      context: String
  ): OffsetDateTime =
    optionalTimestamp(row, primary)
      .orElse(optionalTimestamp(row, secondary))
      .getOrElse(
        throw new IllegalArgumentException(
          s"Missing timestamp: neither '$primary' nor '$secondary' in $context"
        )
      )

  /** Alias: try primary field, fall back to secondary. */
  def stringAlias(
      row: Json,
      primary: String,
      secondary: String,
      context: String
  ): String =
    optionalString(row, primary)
      .orElse(optionalString(row, secondary))
      .getOrElse(
        throw new IllegalArgumentException(
          s"Missing string: neither '$primary' nor '$secondary' in $context"
        )
      )

  /** Alias: try primary long, fall back to secondary. */
  def longAlias(
      row: Json,
      primary: String,
      secondary: String,
      context: String
  ): Long =
    optionalLong(row, primary)
      .orElse(optionalLong(row, secondary))
      .getOrElse(
        throw new IllegalArgumentException(
          s"Missing long: neither '$primary' nor '$secondary' in $context"
        )
      )

  def boolFlag(row: Json, field: String): Long =
    row.hcursor.get[Boolean](field).toOption match
      case Some(true) => 1L
      case _ =>
        row.hcursor.get[Long](field).toOption match
          case Some(v) if v != 0 => 1L
          case _                 => 0L

  def nestedLong(row: Json, parent: String, field: String): Option[Long] =
    for
      parentObj <- row.hcursor.downField(parent).focus
      value     <- parentObj.hcursor.get[Long](field).toOption
    yield value

  def nestedDouble(row: Json, parent: String, field: String): Option[Double] =
    for
      parentObj <- row.hcursor.downField(parent).focus
      value     <- parentObj.hcursor.get[Double](field).toOption
    yield value

  def jsonArrayString(row: Json, field: String): Option[String] =
    row.hcursor.downField(field).focus.flatMap { value =>
      if value.isNull || !value.isArray then None
      else Some(value.noSpaces)
    }

  def rawJson(row: Json): Option[String] =
    if row.isNull then None
    else Some(row.noSpaces)

  def rowSequence(index: Int, context: String): Long =
    if index < 0 then
      throw new IllegalArgumentException(
        s"Negative row sequence index $index in $context"
      )
    index.toLong

  private def parseTimestamp(value: String): Option[OffsetDateTime] =
    Try(OffsetDateTime.parse(value, timestampParser)).toOption
      .orElse(Try(OffsetDateTime.parse(value)).toOption)